File: src/org/algo/排序/QuickSort.java
package org.algo.排序;


public class QuickSort {
    public static int[] quickSort(int[] arr) {
        return quickSort(arr, 0, arr.length - 1);
    }

    public static int[] quickSort(int[] arr, int left, int right) {
        if (left < right) {
            int partitionIndex = partition(arr, left, right);
            // 左半部分递归
            quickSort(arr, left, partitionIndex-1);
            // 右半部分递归
            quickSort(arr, partitionIndex+1, right);
        }
        return arr;
    }

    private static int partition(int[] arr, int left, int right) {
        int pivot = left;
        int index = pivot + 1;
        for (int i = index; i <= right; i++) {
            if (arr[i] < arr[pivot]) {
                swap(arr, i, index);
                index++;
            }
        }
        swap(arr, pivot, index-1);
        return index - 1;
    }

    private static void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    public static void main(String[] args) {
        int[] arr = new int[]{3,4, 2,1, 5};
        int[] res = quickSort(arr);
        for (int num : res) {
            System.out.print(num + " ");
        }
    }
}

File: src/org/algo/排序/golang/quickSort.go
package main

/**
快速排序的主要思想是通过划分将待排序的序列分成前后两部分，其中前一部分的数据都比后一部分的数据要小，然后再递归调用函数对两部分的序列分别进行快速排序，以此使整个序列达到有序。
 */
func quickSort(q []int, l, r int) {
	if l >= r { // 终止条件
		return
	}
	x := q[(l+r)>>1] // 确定分界点
	i, j := l-1, r+1 // 两个指针，因为do while要先自增/自减
	for i < j {      // 每次迭代
		for { // do while 语法
			i++ // 交换后指针要移动，避免没必要的交换
			if q[i] >= x {
				break
			}
		}
		for {
			j--
			if q[j] <= x {
				break
			}
		}
		if i < j { // swap 两个元素
			q[i], q[j] = q[j], q[i]
		}
	}
	quickSort(q, l, j) // 递归处理左右两段
	quickSort(q, j+1, r)
}

